ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java

```diff
@@ -479,10 +495,54 @@ private String visitBooleanComparisonExpression(final ComparisonExpression.Type
       final Pair<String, Schema> right = process(node.getRight(), unmangleNames);
       return new Pair<>(
           "(" + left.getLeft() + " " + node.getType().getValue() + " " + right.getLeft() + ")",
-          Schema.OPTIONAL_FLOAT64_SCHEMA
+          SchemaUtil.resolveArithmeticType(left.getRight().type(), right.getRight().type())
       );
     }
 
+    @Override
+    protected Pair<String, Schema> visitSearchedCaseExpression(
+        final SearchedCaseExpression node,
+        final Boolean unmangleNames) {
+      final String functionClassName = SearchedCaseFunction.class.getSimpleName();
+      final List<CaseWhenProcessed> whenClauses = node
+          .getWhenClauses()
+          .stream()
+          .map(whenClause -> new CaseWhenProcessed(
+              process(whenClause.getOperand(), unmangleNames),
+              process(whenClause.getResult(), unmangleNames)
+          ))
+          .collect(Collectors.toList());
+      final Schema resultSchema = whenClauses.get(0).thenProcessResult.getRight();
+      final String resultSchemaString = SchemaUtil.getJavaType(resultSchema).getCanonicalName();
+
+      final List<String> lazyWhenClause = whenClauses
+          .stream()
+          .map(processedWhenClause -> functionClassName + ".whenClause("
+              + buildSupplierCode(
+              "Boolean", processedWhenClause.whenProcessResult.getLeft())
+              + ", "
+              + buildSupplierCode(
+              resultSchemaString, processedWhenClause.thenProcessResult.getLeft())
+              + ")")
+          .collect(Collectors.toList());
+
+      final String defaultValue = node.getDefaultValue().isPresent()
+          ? process(node.getDefaultValue().get(), unmangleNames).getLeft()
+          : "null";
+
+      final String codeString = "((" + resultSchemaString + ")"
+          + functionClassName + ".searchedCaseFunction(ImmutableList.of( "
+          + StringUtils.join(lazyWhenClause, ", ") + "),"
+          + buildSupplierCode(resultSchemaString, defaultValue)
+          + "))";
+      return new Pair<>(codeString, resultSchema);
+    }
+
+    private String buildSupplierCode(final String typeString, final String code) {
```

Comment:
nit: can be static.

Reply:
This is in the inner class that cannot have static declaration.